Send bokeh plots to the plots pane

seeM opened the pull request:
Addresses #4249.

### QA Notes

We're (unfortunately) inspecting the stack so it's worth testing on different Bokeh versions, Python versions, platforms, etc.

This should open in the plots pane, not the viewer pane:

```python
from bokeh.plotting import figure, show
p = figure(title="Simple line example", x_axis_label='x', y_axis_label='y')
p.line([1, 2, 3, 4, 5], [6, 7, 2, 4, 5], legend_label="Temp.", line_width=2)
show(p)
```

seeM committed:
send bokeh plots to the plots pane

petetronic commented:
Looks like we need to pin different versions based on Python version given test failures.

ERROR: Ignored the following versions that require a different python version: 3.5.0 Requires-Python >=3.10; 3.5.0.dev1 Requires-Python >=3.10; 3.5.0.dev3 Requires-Python >=3.10; 3.5.0.dev4 Requires-Python >=3.10; 3.5.0.dev5 Requires-Python >=3.10; 3.5.0.dev6 Requires-Python >=3.10; 3.5.0.dev7 Requires-Python >=3.10; 3.5.0.dev8 Requires-Python >=3.10; 3.5.0rc1 Requires-Python >=3.10; 3.5.0rc2 Requires-Python >=3.10; 3.5.1 Requires-Python >=3.10; 3.6.0.dev1 Requires-Python >=3.10
ERROR: Could not find a version that satisfies the requirement bokeh==3.5.1 (from versions: 0.2, 0.3, 0.4, 0.4.1, 0.4.4, 0.5.0, 0.5.1, 0.5.2, 0.6.0, 0.6.1, 0.7.0, 0.7.1, 0.8.0, 0.8.1, 0.8.2, 0.9.0, 0.9.1, 0.9.2, 0.9.3, 0.10.0, 0.11.0, 0.11.1, 0.12.0, 0.12.1, 0.12.2, 0.12.3, 0.12.4, 0.12.5, 0.12.6, 0.12.7, 0.12.9, 0.12.10, 0.12.11, 0.12.13, 0.12.14, 0.12.15, 0.12.16, 0.13.0, 1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.0.4, 1.1.0, 1.2.0, 1.3.0, 1.3.1, 1.3.2, 1.3.4, 1.4.0, 2.0.0rc1, 2.0.0rc2, 2.0.0, 2.0.1, 2.0.2, 2.1.0rc1, 2.1.0rc2, 2.1.0rc3, 2.1.0, 2.1.1, 2.2.0.dev4, 2.2.0.dev6, 2.2.0.dev7, 2.2.0.dev8, 2.2.0rc1, 2.2.0rc2, 2.2.0rc3, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.3.0.dev1, 2.3.0.dev2, 2.3.0.dev3, 2.3.0.dev5, 2.3.0.dev7, 2.3.0.dev8, 2.3.0.dev9, 2.3.0.dev10, 2.3.0.dev11, 2.3.0.dev12, 2.3.0.dev13, 2.3.0.dev14, 2.3.0rc2, 2.3.0rc3, 2.3.0rc4, 2.3.0, 2.3.1.dev1, 2.3.1rc1, 2.3.1rc2, 2.3.1, 2.3.2rc1, 2.3.2, 2.3.3rc1, 2.3.3, 2.4.0.dev1, 2.4.0.dev2, 2.4.0.dev4, 2.4.0.dev5, 2.4.0.dev6, 2.4.0.dev7, 2.4.0.dev8, 2.4.0rc1, 2.4.0rc2, 2.4.0rc3, 2.4.0, 2.4.1rc1, 2.4.1, 2.4.2.dev1, 2.4.2rc1, 2.4.2, 2.4.3.dev1, 2.4.3.dev3, 2.4.3.dev4, 2.4.3rc1, 2.4.3, 3.0.0.dev1, 3.0.0.dev2, 3.0.0.dev3, 3.0.0.dev4, 3.0.0.dev5, 3.0.0.dev6, 3.0.0.dev7, 3.0.0.dev8, 3.0.0.dev9, 3.0.0.dev10, 3.0.0.dev11, 3.0.0.dev13, 3.0.0.dev14, 3.0.0.dev15, 3.0.0.dev16, 3.0.0.dev17, 3.0.0.dev18, 3.0.0.dev19, 3.0.0.dev20, 3.0.0rc1, 3.0.0rc2, 3.0.0rc3, 3.0.0rc4, 3.0.0rc5, 3.0.0rc6, 3.0.0, 3.0.1rc1, 3.0.1, 3.0.2rc1, 3.0.2, 3.0.3rc1, 3.0.3, 3.1.0.dev1, 3.1.0.dev2, 3.1.0.dev3, 3.1.0.dev4, 3.1.0.dev5, 3.1.0rc1, 3.1.0rc2, 3.1.0rc3, 3.1.0, 3.1.1rc1, 3.1.1, 3.2.0.dev1, 3.2.0.dev2, 3.2.0.dev3, 3.2.0.dev4, 3.2.0rc1, 3.2.0rc2, 3.2.0rc3, 3.2.0, 3.2.1, 3.2.2rc1, 3.2.2, 3.3.0.dev1, 3.3.0.dev2, 3.3.0.dev3, 3.3.0rc1, 3.3.0rc2, 3.3.0rc3, 3.3.0, 3.3.1.dev1, 3.3.1rc1, 3.3.1, 3.3.2, 3.3.3, 3.3.4rc1, 3.3.4, 3.4.0.dev1, 3.4.0.dev2, 3.4.0.dev3, 3.4.0.dev4, 3.4.0.dev5, 3.4.0.dev6, 3.4.0.dev7, 3.4.0.dev8, 3.4.0rc1, 3.4.0rc2, 3.4.0, 3.4.1, 3.4.2, 3.4.3)
ERROR: No matching distribution found for bokeh==3.5.1

petetronic committed:
Pin bokeh test dependency based on python version

Python 3.9 support was dropped in Bokeh 3.5.0
Python 3.8 support was dropped in Bokeh 3.2.0

nstrayer commented on extensions/positron-python/python_files/positron/positron_ipykernel/ui.py:
```diff
@@ -181,13 +184,31 @@ def open(self, url, new=0, autoraise=True):
+            # Identify bokeh plots by checking the stack for the bokeh.io.showing.show function.
+            # This is not great but currently the only information we have.
+            bokeh_io_showing = sys.modules.get("bokeh.io.showing")
+            if bokeh_io_showing:
+                for frame_info in inspect.stack():
+                    if (
+                        inspect.getmodule(frame_info.frame, frame_info.filename) == bokeh_io_showing
+                        and frame_info.function == "show"
+                    ):
+                        is_plot = True
+                        break
```
It's premature at this point but I wonder if it makes sense to have a registry of things like this we can add to in the future that takes the current stack and other appropriate inputs. I can't imagine this is the last time we'll be touching this type of thing.

seeM replied:
Yeah we may end up with something like that. I'm hoping there are more direct and robust ways to integrate with other libraries, but we'll see!

nstrayer commented on extensions/positron-python/python_files/positron/positron_ipykernel/ui.py:
```diff
@@ -181,13 +184,31 @@ def open(self, url, new=0, autoraise=True):
-                    # TODO: Figure out if the file being displayed is a plot or not.
-                    "is_plot": False,
-                },
+                payload=ShowHtmlFileParams(
```
Oh good fix with the `ShowHtmlFileParams()`, I didn't realize there was a `.dict()` method!

petetronic replied:
Is .dict() deprecated in Pydantic? Is the new equivalent `model_dump()`?

petetronic replied:
(I think we're on a much older version of Pydantic, so it looks like dict() is valid. Later versions v2+ have replaced that with model_dump() it seems).

nstrayer reviewed (dismissed):
Looks great! I'm excited to finally have some sort of answer / pattern for the existential question of "how do we know what a plot is?"

petetronic commented:
Looks like we'll need to dig into the test for Bokeh on Windows

seeM committed:
fix bokeh plots on windows

seeM committed:
add test debug config to step into external code

seeM commented:
I've fixed the Windows bug in [2227e58](https://github.com/posit-dev/positron/pull/4331/commits/2227e58f5e7a43407d29f5a45aaa62e665a924d2). I think that may also be a bug on `main`.

seeM committed:
remove duplicate code + formatter
